src/main/java/com/xpn/xwiki/plugin/lucene/searcherProvider/SearcherProvider.java

```diff
@@ -156,17 +149,14 @@ protected void finalize() throws Throwable {
   }
 
   public void connectSearchResults(SearchResults searchResults) {
-    if (!this.checkConnected()) {
-      throw new IllegalStateException("you may not connect a searchResult to a"
-          + " SearchProvider from a not connected thread.");
-    }
+    checkState(checkConnected(), "you may not connect a searchResult to a SearchProvider from"
+        + " a not connected thread.");
     getConnectedSearchResultsForCurrentThread().add(searchResults);
   }
 
   Set<SearchResults> getConnectedSearchResultsForCurrentThread() {
-    if (!hasSearchResultsForCurrentThread()) {
-      connectedSearchResultsMap.put(Thread.currentThread(), new HashSet<SearchResults>());
-    }
+    connectedSearchResultsMap.putIfAbsent(Thread.currentThread(), Collections.newSetFromMap(
+        new ConcurrentHashMap<SearchResults, Boolean>()));
```

Comment:
I think the inserted set can be a regular HashSet since it is only ever accessed by one thread.